### PR TITLE
Add method to enumerate zfs props.

### PR DESCRIPTION
**Pull request by jgrund**
Add a method to enumerate zfs props from zfs_props.

A method to enumerate zpool props will be added later if it's needed.

  - Add method to enumerate zfs props using underlying mechanism of `zfs get all`.
  - Split rust-libzfs into discrete module files.
  - Update num cores in vagrant box to test parallelism.

Signed-off-by: Joe Grund <joe.grund@intel.com>

### CODE REVIEW

File: libzfs-sys/src/lib.rs [hunk 1]
```diff
@@ -241,31 +265,25 @@ mod tests {
...
-            }
-
             while !pl_p.is_null() {
                 let zfs_prop = to_zfs_prop_t((*pl_p).pl_prop).unwrap();
 
                 if zfs_prop != zfs_prop_t::ZFS_PROP_BAD {
-                    let raw = create_with_capacity(ZFS_MAXPROPLEN);
-                    let source_raw = create_with_capacity(ZFS_MAX_DATASET_NAME_LEN);
-                    let source_t = ptr::null_mut();
+                    let raw = CString::new("0".repeat(319)).unwrap().into_raw();
```
Comment (johnsonw) [inline libzfs-sys/src/lib.rs]: What is 319?

File: libzfs/src/vdev.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+    fn lookup_is_log(tree: &nvpair::NvList) -> Option<bool> {
+        tree.lookup_uint64(sys::zpool_config_is_log())
+            .map(|x| x == 1)
+            .ok()
+    }
+
+    fn lookup_guid(tree: &nvpair::NvList) -> Option<String> {
+        tree.lookup_uint64(sys::zpool_config_guid())
+            .map(|x| format!("{:#018X}", x))
```
Comment (johnsonw) [inline libzfs/src/vdev.rs]: What is #018X?

File: libzfs/src/vdev.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        state.to_owned().into_string().map_err(LibZfsError::from)
+    }
+
+    match x {
+        x if x == sys::VDEV_TYPE_DISK => {
+            let path = tree.lookup_string(sys::zpool_config_path())?.into_string()?;
+            let dev_id = lookup_tree_str(tree, sys::zpool_config_dev_id())?;
+            let phys_path = lookup_tree_str(tree, sys::zpool_config_phys_path())?;
+            let whole_disk = tree.lookup_uint64(sys::zpool_config_whole_disk())
+                .map(|x| x == 1)
```
Comment (johnsonw) [inline libzfs/src/vdev.rs]: Does this represent that whole disk is true? If so, didn't you put in a patch for boolean types where 1 represents true? Should this be replaced with that type?

File: libzfs/src/zfs.rs [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+            )
+        };
+
+        match code {
+            0 => Ok(ZpropList::new(prop_list_ptr)),
+            x => Err(LibZfsError::Io(Error::from_raw_os_error(x))),
+        }
+    }
+    pub fn props(&self) -> Result<(Vec<ZProp>)> {
+        let buff_size = 319;
```
Comment (johnsonw) [inline libzfs/src/zfs.rs]: What's the significance of 319?

File: libzfs-sys/src/lib.rs (hunk 1 above)
Comment (jgrund) [reply libzfs-sys/src/lib.rs]: 320 is the maximum size I found usable for the pre-allocated buffer that gets the property.

319 + null byte = 320.

File: libzfs/src/zfs.rs (hunk 4 above)
Comment (jgrund) [reply libzfs/src/zfs.rs]: 320 is the maximum size I found usable for the pre-allocated buffer that gets the property.

319 + null byte = 320.

File: libzfs/src/vdev.rs (hunk 2 above)
Comment (jgrund) [reply libzfs/src/vdev.rs]: It represents padding width + hex formatting. Here is the patch that it was changed in: https://github.com/intel-hpdd/rust-libzfs/pull/33

File: libzfs/src/vdev.rs (hunk 3 above)
Comment (jgrund) [reply libzfs/src/vdev.rs]: > Does this represent that whole disk is true? 

Yes.

>  If so, didn't you put in a patch for boolean types where 1 represents true? 

No, the generated zfs bindings have a type where 0 == false & 1 == true.

> Should this be replaced with that type?

I don't think so. The ZFS source usages uses 0 & 1 for comparison with this result.

**Pull request merged**